revokefs/writer.c

```diff
@@ -0,0 +1,909 @@
+      return 0;
+    }
+
+  r = fsync (fd);
+  if (r == -1)
+    response->result = -errno;
+  else
+    response->result = r;
+
+  return 0;
+}
+
+int
+request_fsync (int writer_socket, int fd)
+{
+  RevokefsRequest request = { REVOKE_FS_FSYNC };
+  RevokefsResponse response;
+  ssize_t response_data_len;
+
+  request.arg1 = fd;
+
+  response_data_len = do_request (writer_socket, &request, NULL, 0, NULL, 0,
+                                  &response, NULL, 0);
+  if (response_data_len < 0)
+    return -EIO;
+
+  return response.result;
+}
+
+static ssize_t
+handle_close (RevokefsRequest *request,
+              gsize data_size,
+              RevokefsResponse *response)
+{
+  int r;
+  int fd = request->arg1;
+  size_t size = request->arg2;
+  off_t offset = request->arg3;
+
+  if (!g_hash_table_remove (outstanding_fds, GUINT_TO_POINTER(fd)))
+    {
+      response->result = -EBADFD;
+      return 0;
+    }
+
+  close (fd);
+  response->result = 0;
+  return 0;
+}
+
+int
+request_close (int writer_socket, int fd)
+{
+  RevokefsRequest request = { REVOKE_FS_CLOSE };
+  RevokefsResponse response;
+  ssize_t response_data_len;
+
+  request.arg1 = fd;
+  response_data_len = do_request (writer_socket, &request, NULL, 0, NULL, 0,
+                                  &response, NULL, 0);
+  if (response_data_len < 0)
+    return -EIO;
+
+  return response.result;
+}
+
+static ssize_t
+handle_access (RevokefsRequest *request,
+               gsize data_size,
+               RevokefsResponse *response)
+{
+  g_autofree char *path = get_valid_path (request->data, data_size);
+  int mode = request->arg1;
+
+  /* Apparently at least GNU coreutils rm calls `faccessat(W_OK)`
+   * before trying to do an unlink.  So...we'll just lie about
+   * writable access here.
+   */
+  if (faccessat (basefd, path, mode, AT_SYMLINK_NOFOLLOW) == -1)
+    response->result = -errno;
+  else
+    response->result = 0;
+
+  return 0;
+}
+
+int
+request_access (int writer_socket, const char *path, int mode)
+{
+  return request_path_int (writer_socket, REVOKE_FS_ACCESS, path, mode);
+}
+
+static gpointer
+monitor_exit_fd_thread (gpointer user_data)
+{
+  guchar eof;
+  int exit_with_fd = GPOINTER_TO_INT (user_data);
+
+  while (1)
+    {
+      if (read (exit_with_fd, &eof, sizeof (eof)) == 0)
+        {
+          perror ("Received EOF on exit-with-fd argument");
+          exit (1);
+        }
+    }
+
+  return NULL;
+}
+
+void
+do_writer (int basefd_arg,
+           int fuse_socket,
+           int exit_with_fd)
+{
+  guchar request_buffer[MAX_REQUEST_SIZE];
+  RevokefsRequest *request = (RevokefsRequest *)&request_buffer;
+  guchar response_buffer[MAX_RESPONSE_SIZE];
+  RevokefsResponse *response = (RevokefsResponse *)&response_buffer;
+
+  if (exit_with_fd)
+    {
+      g_autoptr (GThread) monitor_exit_fd = NULL;
+      g_autoptr (GError) error = NULL;
+
+      monitor_exit_fd = g_thread_try_new (NULL, monitor_exit_fd_thread, GINT_TO_POINTER (exit_with_fd), &error);
```

Comment:
What's the purpose of using `g_thread_try_new` rather than `g_thread_new`? The only difference between the two is that the latter aborts on error; but that's essentially what we're doing here.